### PR TITLE
DataManager used for cell measures

### PR DESCRIPTION
**Pull request by pp-mo**
TODO:
 * [x] check error when creating cell measures with lazy integer data
 * [x] check error when saving masked cell measures data to netcdf

### CODE REVIEW

File: lib/iris/coords.py [hunk 1]
```diff
@@ -1762,30 +1761,42 @@ def data(self, data):
         if data is None:
             raise ValueError('The data payload of a CellMeasure may not be '
                              'None; it must be a numpy array or equivalent.')
+        if is_lazy_data(data) and data.dtype.kind in 'iu':
```
Comment (dkillick) [inline lib/iris/coords.py]: Let's stretch a crazy thing a little further: do you think we might need to cope with boolean cell measures? We have already discussed (offline) the concept of a cell measure containing only 1 or 0, which is effectively a boolean cell measure...

File: lib/iris/coords.py [hunk 2]
```diff
@@ -1762,30 +1761,42 @@ def data(self, data):
...
+            # Disallow lazy integer data.  Integer values would be very
+            # unexpected anyway, but this avoids the *even more obscure*
+            # possibility that there are masked values, because masked integer
+            # data would cause dask to return data with the wrong dtype.
+            # This can all be fixed, as for cube data, by adding fill_value and
+            # dtype support, but for now we prefer just not to support this.
```
Comment (dkillick) [inline lib/iris/coords.py]: And we can add this as a feature request at a later date once we have a concrete example of this being needed :+1:

File: lib/iris/fileformats/netcdf.py [hunk 3]
```diff
@@ -1494,9 +1494,18 @@ def _create_cf_cell_measure_variable(self, cube, dimension_names,
...
                          cube.cell_measure_dims(cell_measure)]
 
+        # Get the data values
+        data = cell_measure.data
+
+        # Disallow saving of *masked* cell measures.
+        if ma.is_masked(data):
+            # We can't really save properly if there are masked points, because
+            # we don't (yet) record an appropriate fill_value property.
+            msg = "Currently cannot save cell measures with missing data."
```
Comment (dkillick) [inline lib/iris/fileformats/netcdf.py]: I'd drop the "currently" -- we _just don't support this_, and that's just how it is unless someone asks for it.

File: lib/iris/fileformats/netcdf.py [hunk 4]
```diff
@@ -1494,9 +1494,18 @@ def _create_cf_cell_measure_variable(self, cube, dimension_names,
         cf_dimensions = [dimension_names[dim] for dim in
                          cube.cell_measure_dims(cell_measure)]
 
+        # Get the data values
```
Comment (dkillick) [inline lib/iris/fileformats/netcdf.py]: Strictly you should add a full stop at the end of this comment.

File: lib/iris/coords.py [hunk 5]
```diff
@@ -1762,30 +1761,42 @@ def data(self, data):
...
             data = np.array(data, ndmin=1)
-        if hasattr(self, '_data') and self._data is not None:
+        if hasattr(self, '_data_manager') and self._data_manager is not None:
             # Check that setting these data wouldn't change self.shape
             if data.shape != self.shape:
                 raise ValueError("New data shape must match existing data "
                                  "shape.")
 
-        self._data = data
+        self._data_manager = DataManager(data)
```
Comment (dkillick) [inline lib/iris/coords.py]: Do you think there's value in a comment here explaining why we don't create a `DataManager` until this point? Might not be clear to someone fresh to the code.

File: lib/iris/coords.py [hunk 6]
```diff
@@ -1762,30 +1761,42 @@ def data(self, data):
...
 
     @property
     def shape(self):
         """The fundamental shape of the Cell Measure, expressed as a tuple."""
-        # Access the underlying _data attribute to avoid triggering
+        # Access the underlying _data_manager attribute to avoid triggering
```
Comment (dkillick) [inline lib/iris/coords.py]: We could probably drop this comment entirely now, right?

File: lib/iris/coords.py (hunk 1 above)
Comment (pp-mo) [inline lib/iris/coords.py]: This is still all out of scope as-of CF v1.7, but here we are covering ourselves for whatever might exist in a file, 
So yes, I'll add that as another.

File: lib/iris/coords.py (hunk 2 above)
Comment (pp-mo) [inline lib/iris/coords.py]: I think I'll rewrite to make that clearer...

File: lib/iris/coords.py (hunk 5 above)
Comment (pp-mo) [inline lib/iris/coords.py]: I'll add comments in all of `__init__`, `copy` and `__getitem__` to make clear how it all fits together.

File: lib/iris/coords.py (hunk 6 above)
Comment (pp-mo) [inline lib/iris/coords.py]: Yes

**Comment by pp-mo** [comment]
> e6a8045 "Review changes: clearer comments."

Hopefully addresses all you asked for, plus a couple of other improvements.

**Comment by pp-mo** [comment]
Still to-do : https://github.com/pp-mo/iris/pull/43

**Comment by dkillick** [comment]
@pp-mo I'm :+1: with the changes to this point. Now all we need to do is integrate the outstanding testing and I think we're there.

**Comment by pp-mo** [comment]
> all we need to do is integrate the outstanding testing 

Now done.
Ready to go ?

File: lib/iris/tests/unit/coords/test_CellMeasure.py [hunk 7]
```diff
@@ -100,6 +100,14 @@ def test___getitem__(self):
         sub_measure = self.measure[2]
         self.assertArrayEqual(self.values[2], sub_measure.data)
 
+    def test___getitem__data_copy(self):
+        # Check that a sliced cell measure has independent data.
+        sub_measure = self.measure[1:3]
+        old_values = sub_measure.data.copy()
+        self.measure.data[:] = 0.0
+        new_values = sub_measure.data.copy()
```
Comment (dkillick) [inline lib/iris/tests/unit/coords/test_CellMeasure.py]: Is this second copy call necessary?

File: lib/iris/coords.py [hunk 8]
```diff
@@ -1813,11 +1813,16 @@ def __getitem__(self, key):
         # each dimension of the cell_measure.
         full_slice = iris.util._build_full_slice_given_keys(key, self.ndim)
 
+        # Get the data, all or part of which will become the new data.
+        data = self._data_manager.core_data.copy()
+        # Copy the data to avoid making the new measure a view on the old one.
+        data = data.copy()
```
Comment (dkillick) [inline lib/iris/coords.py]: You've now copied this variable _twice_. That seems like a lot of copying...

**Comment by dkillick** [comment]
@pp-mo that will do nicely, thanks!

File: lib/iris/coords.py [hunk 9]
```diff
@@ -1804,18 +1813,19 @@ def __getitem__(self, key):
         # each dimension of the cell_measure.
         full_slice = iris.util._build_full_slice_given_keys(key, self.ndim)
 
+        # Get the data, all or part of which will become the new data.
+        data = self._data_manager.core_data
+        # Copy the data to avoid making the new measure a view on the old one.
+        data = data.copy()
```
Comment (pp-mo) [inline lib/iris/coords.py]: This line added to prevent views.

**Pull request merged**